contracts/zero-ex/contracts/src/features/OtcOrdersFeature.sol

```diff
@@ -0,0 +1,467 @@
+                !stor.orderSignerRegistry[order.maker][makerSigner]
+            ) {
+                LibNativeOrdersRichErrors.OrderNotSignedByMakerError(
+                    orderInfo.orderHash,
+                    makerSigner,
+                    order.maker
+                ).rrevert();
+            }
+
+            // If msg.sender is not the taker, validate the taker signature.
+            if (!_isSenderValidTaker(order.taker)) {
+                address takerSigner = LibSignature.getSignerOfHash(orderInfo.orderHash, takerSignature);
+                if (
+                    takerSigner != order.taker &&
+                    !stor.orderSignerRegistry[order.taker][takerSigner]
+                ) {
+                    LibNativeOrdersRichErrors.OrderNotSignedByTakerError(
+                        orderInfo.orderHash,
+                        takerSigner,
+                        order.taker
+                    ).rrevert();
+                }
+                taker = order.taker;
+            }            
+        }
+
+        // Settle between the maker and taker.
+        (takerTokenFilledAmount, makerTokenFilledAmount) = _settleOtcOrder(
+            order,
+            taker,
+            takerTokenFillAmount,
+            wethOptions
+        );
+
+        emit OtcOrderFilled(
+            orderInfo.orderHash,
+            order.maker,
+            taker,
+            address(order.makerToken),
+            address(order.takerToken),
+            takerTokenFilledAmount,
+            makerTokenFilledAmount
+        );
+    }
+
+    /// @dev Settle the trade between an OTC order's maker and taker.
+    /// @param order The OTC order.
+    /// @param takerTokenFillAmount Maximum taker token amount to fill this
+    ///        order with.
+    /// @return takerTokenFilledAmount How much taker token was filled.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function _settleOtcOrder(
+        LibNativeOrder.OtcOrder memory order,
+        address taker,
+        uint128 takerTokenFillAmount,
+        WethOptions wethOptions
+    )
+        private
+        returns (uint128 takerTokenFilledAmount, uint128 makerTokenFilledAmount)
+    {
+        {
+            // Unpack nonce fields
+            uint64 nonceBucket = uint64(order.expiryAndNonce >> 128);
+            uint128 nonce = uint128(order.expiryAndNonce);
+            // Update tx origin nonce for the order
+            LibOtcOrdersStorage.getStorage().txOriginNonces
+                [order.txOrigin][nonceBucket] = nonce;
+        }
+
+        // Clamp the taker token fill amount to the fillable amount.
+        takerTokenFilledAmount = LibSafeMathV06.min128(
+            takerTokenFillAmount,
+            order.takerAmount
+        );
+        // Compute the maker token amount.
+        // This should never overflow because the values are all clamped to
+        // (2^128-1).
+        makerTokenFilledAmount = uint128(LibMathV06.getPartialAmountFloor(
```

Comment:
curious if this calculation is expensive at all, for taker signed flow we know `makerTokenFilledAmount` = `order.makerAmount` if we can identify it at this point

Reply:
good call 👍